text/ProofModes.md

```diff
@@ -0,0 +1,118 @@
+   mutable and having specific commands assign such non-terminal to the desired one.
+   The current proof mode is not part of the `Summary` and is both stored in `Proof_global`
+   and `Stm` that do their best to ensure parsing happens in a state where such non-terminal
+   holds the expected value.
+
+## Scenario not supported by the status quo
+
+1. using `TACTIC EXTEND` for a non-terminal that is not ltac specific.
+
+2. loading two proof languages with syntactic conflicts at the *grammar* level
+   and using both.
+   
+    In `mode1` we have `Tactic Notation "foo" "bar" constr(x) := idtac x.`
+    In `mode2` we have a conflicting grammar `Tactic Notation "foo" constr(x) := idtac x.`
+    If we load both
+    
+    ```    
+    Goal nat -> True.
+    intro bar.
+    foo bar bar.  (* works *)
+    foo bar. (* does not parse *)
+    ```
+    
+    Of course the grammar is ambiguous, but if the two modes were not blindly mixed
+    one could imagine one of the following to work:
+    
+    `mode2:: rewrite ->H => [].`
+    `mode1.(foo => bar).`
+    
+    ```
+    begin mode2.
+      foo bar.
+    end.
+    ```
+    
+    Where `mode2::` and `begin mode2` are understood by the parsing engine.
+
+2. loading two proof languages with syntactic conflicts at the *lexer* level
+   and using both.
+
+   In `mode1` we have `Tactic Notation "foo" "=>" ">" := idtac 1.`
+   and in `mode2` we have `Tactic Notation "bar" "=>>" := idtac 2.`
+   If only `mode1` is loaded the sentence `foo =>>` parses fine,
+   while it stops working when the grammar rule in `mode2` declares
+   `=>>` as a single token.
+   
+   Another similar problem used to arise between IDENT if one declares a `Notation "'foo'"` for a
+   term, fixed in 982460743a54ecfab1d601ba930d61c04972d17a 
+
+1. Decouple the proof language from the `.vo` file.  Since the `.vo` contains the `Libobject`
+   action corresponding to `Declare ML Module` one gets both (even by transitivity).  It should be
+   possible to access a theorem without caring about the proof mode used to prove it.
+
+
+## Proposal
+
+Each proof mode installs all its grammar in a dedicated non-terminal `mode-root`.
+Such non-terminal is installed as the main entry (today the one for `tactic-expr`)
+in one of the following ways:
+- `Set Default Proof Mode "mode1".` (TODO: disable inside proof?)
+- `begin mode1. .... end.` within a proof.
+- `mode1:: ... .` within a proof, changes mode for 1 statement only
+- `Proof .... mode mode1 ... using ... .` (Useful?)
+
+Also fix the token gluing problem (proposal https://coq.inria.fr/bugs/show_bug.cgi?id=4867 )
+to make it simpler to reuse non-terminals of other modes inside the grammar of a
+mode if one likes to.
+
+## Bugs
+
+In this case we don't get why the second rule does not mask the first one.
+
+```
+Notation "! 'bar' x " := 1 (at level 2).
+Notation "! bar x" := x (at level 2).
+
+Goal ! bar 2 = 3.
+```
+
```

Comment:
`Print Grammar constr` shows:
```
...
| "2" RIGHTA
  [ "!"; IDENT "bar"; NEXT
  | "!"; constr:operconstr LEVEL "200"; NEXT ]
...
```
This indicates that both rules are active (and overlapping only on expressions of the form `! bar any-term`). It seems that the convention of camlp5 is to give priority to the rule which parses a terminal, so here trying to parse the terminal `IDENT "bar"` before trying to parse the non terminal `operconstr`.